app/uk/gov/hmrc/helptosavereminder/util/DateTimeFunctions.scala

```diff
@@ -30,7 +30,7 @@ object DateTimeFunctions {
 
     nextAvailableDayOfMonth match {
       case Some(day) => (LocalDate.now).plusDays(day - currentDayOfMonth)
-      case None      => (LocalDate.now).plusMonths(1).withDayOfMonth(validDaysToReceive.head)
+      case None      => (LocalDate.now).plusMonths(1).withDayOfMonth(validDaysToReceive.headOption.getOrElse(1))
```

Comment:
This doesn't eve do the right thing!  validDaysToReceive is validated by the current month, but then you go to the next month but only accept it if the day was in the month before.  If all else fails, you chose 1 arbitrarily

Reply:
Its actually won't do anything meaningful except to pass the wartremover test. Can you let me know what can be the default thing to set up?

Reply:
Yeah, I think I can go with nextAvailableDayOfMonth as the default day instead of 1.

Reply:
The old code was already wrong, as I say it filters according to the current month but tries to picks that day in the next month, even if the next month does not have that day.

Part of the problem is that this pretty much has to be runtime, because we can't represent the constraints in the types of the objects.  That means that one way or another we have to throw when we cannot return a correct value.

Anyway, I think this needs rewriting because it is not testable.  It is not testable because it use Calendar.getInstance and LocalDate.now, so will never produce testable results.

Sorry, but the more I look at this the more disappointing I find this code.  I would expect better.

------

Where shall I start?

The code is not safe because it uses Calendar.getInstance and LocalDate.now in various places and there is no guarantee that the date will not change between one call and another.  Nowhere is this more obvious than in:
```
YearMonth
      .of(Calendar.getInstance.get(Calendar.YEAR), Calendar.getInstance.get(Calendar.MONTH) + 1))
```
depending on the order in which the expression gets evaluated, this can be one year out.

Of course, those things will only go wrong rarely, but it is not difficult to write code that clearly can work all of the time

Apart from that, is it the function called getMaxDaysInMonth actually getting the days in next month (because of the "+ 1")?

I could go on

------

Please write me a new class or object that on can have some confidence in.  I would suggest that it operates starting from a single LocalDateTime value supplied.  When used, you can call it with a LocalDateTime.now(ZoneId.of("Europe/London"))) and all calendar calculations etc. can be based on that

Reply:
Actually I ad already had an alternative way to schedule the things. I did the scheduling using Cron Scheduler that expects a regular expression like this: 0 0-10 9,10,11,13,14,15,16,17 1,10,25 * ? *
which is interpreted as (minutes, hours, day of the months, month)
This will help us remove all the code related to these DateTimeFunctions and also we get the advantage of very stable production and time-tested Cron scheduler. I also saw that it avoids the need to wrry about BST/GMT etc.
I think that will help us. If you are okay I will introduce Cron scheduler.

Reply:
OK, which Cron Scheduler is that?

Reply:
This is the Scheduler:
https://github.com/hmrc/group-sync-orchestrator/blob/master/project/AppDependencies.scala
This is how the Job is created:
https://github.com/hmrc/group-sync-orchestrator/blob/master/app/scheduler/ScheduledJob.scala